### PR TITLE
Support Plugin SPI

### PR DESCRIPTION
**Pull request by QilongZhang**
#186

### CODE REVIEW

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/extension/ExtensionLoaderServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+                                .getExtension().value());
+                            if (old == null || old.getPriority() > extensionClass.getPriority()) {
+                                extensionClassMap.put(extensionClass.getExtension().value(),
+                                    extensionClass);
+                            }
+                        }
+                        EXTENSION_MAP.put(interfaceType, extensionClassMap);
+                    } catch (Throwable throwable) {
+                        LOGGER.error("Loading extension of interfaceType: {} occurs error.",
+                            interfaceType);
```
Comment (straybirdzls) [inline sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/extension/ExtensionLoaderServiceImpl.java]: Should log throwable here.

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/extension/ExtensionLoaderServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+import static com.alipay.sofa.ark.spi.constant.Constants.EXTENSION_FILE_DIR;
+
+/**
+ * @author qilong.zql
+ * @since 0.6.0
+ */
+@Singleton
+public class ExtensionLoaderServiceImpl implements ExtensionLoaderService {
+    private static final ConcurrentHashMap<Class, ConcurrentHashMap<String, ExtensionClass>> EXTENSION_MAP = new ConcurrentHashMap<>();
```
Comment (straybirdzls) [inline sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/extension/ExtensionLoaderServiceImpl.java]: Just a suggestion: should we store plugin name here? when use telnet to show plugin info ,can should extension being loaded as well.

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/pipeline/ExtensionLoaderStage.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+@Singleton
+public class ExtensionLoaderStage implements PipelineStage {
+
+    @Inject
+    private ExtensionLoaderService extensionLoaderService;
+
+    @Override
+    public void process(PipelineContext pipelineContext) throws ArkException {
+        ArkServiceLoader.setExtensionLoaderService(extensionLoaderService);
```
Comment (straybirdzls) [inline sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/pipeline/ExtensionLoaderStage.java]: Can add some logs to show all the extensions being loaded.

Comment (QilongZhang) [reply sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/pipeline/ExtensionLoaderStage.java]: @straybirdzls  extension loading  can be trigger any time. and the loading message would be logged with debug level.

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/extension/ExtensionLoaderServiceImpl.java (hunk 1 above)
Comment (QilongZhang) [reply sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/extension/ExtensionLoaderServiceImpl.java]: fixed

File: sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/service/extension/ExtensionClass.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    public I getObject() {
+        if (extensible.singleton()) {
+            return getSingleton();
+        } else {
+            try {
+                return implementClass.newInstance();
+            } catch (Throwable throwable) {
+                throw new ArkException(String.format("Create %s instance error.",
+                    implementClass.getCanonicalName()), throwable);
+            }
```
[the remark this answers is not in the document]
Comment (QilongZhang) [reply sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/service/extension/ExtensionClass.java]: fixed.

**Pull request merged**